MidiApple.cpp: error: cannot bind non-const lvalue reference of type 'void*&' to an rvalue of type 'void*'

This is 1.2.2 (since it is the last one to support Qt4), building with gcc 12.3.0:

```
:info:build [ 56%] Building CXX object src/CMakeFiles/lmmsobjs.dir/core/midi/MidiPort.cpp.o
:info:build cd /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src && /opt/local/bin/g++-mp-12 -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NO_DEBUG -DQT_XML_LIB -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/lmmsobjs_autogen/include -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2 -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/include -I/opt/local/include/opus -I/opt/local/include/SDL -I/opt/local/include -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/3rdparty/rpmalloc/rpmalloc/rpmalloc -isystem /opt/local/libexec/qt4/include -isystem /opt/local/libexec/qt4/include/QtGui -isystem /opt/local/libexec/qt4/include/QtXml -isystem /opt/local/libexec/qt4/include/QtCore -fno-exceptions -Wall -Werror=unused-function -Wno-sign-compare -Wno-strict-overflow -Wno-array-bounds -Wno-attributes -pipe -Os -fPIC -DPIC -std=c++0x -DNDEBUG -arch ppc -mmacosx-version-min=10.6   -D'QT_TRANSLATIONS_DIR="/opt/local/libexec/qt4/share/translations"' -D'LIB_DIR="../lib/"' -D'PLUGIN_DIR="../lib/lmms/"' -MD -MT src/CMakeFiles/lmmsobjs.dir/core/midi/MidiPort.cpp.o -MF CMakeFiles/lmmsobjs.dir/core/midi/MidiPort.cpp.o.d -o CMakeFiles/lmmsobjs.dir/core/midi/MidiPort.cpp.o -c /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiPort.cpp
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp: In member function 'virtual QString MidiApple::sourcePortName(const MidiEvent&) const':
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp:157:41: warning: format '%s' expects argument of type 'char*', but argument 2 has type 'const void*' [-Wformat=]
:info:build   157 |         qDebug("sourcePortName return '%s'?\n", event.sourcePort());
:info:build       |                                        ~^       ~~~~~~~~~~~~~~~~~~
:info:build       |                                         |                       |
:info:build       |                                         char*                   const void*
:info:build       |                                        %p
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp: In member function 'void MidiApple::openDevices()':
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp:442:45: error: cannot bind non-const lvalue reference of type 'void*&' to an rvalue of type 'void*'
:info:build   442 |                 char * deviceName = getName(device);
:info:build       |                                             ^~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp:401:31: note:   initializing argument 1 of 'char* getName(void*&)'
:info:build   401 | char *getName( MIDIObjectRef &object )
:info:build       |                ~~~~~~~~~~~~~~~^~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp:458:56: error: cannot bind non-const lvalue reference of type 'void*&' to an rvalue of type 'void*'
:info:build   458 |                         qDebug("  Entity: %s", getName(entity));
:info:build       |                                                        ^~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp:401:31: note:   initializing argument 1 of 'char* getName(void*&)'
:info:build   401 | char *getName( MIDIObjectRef &object )
:info:build       |                ~~~~~~~~~~~~~~~^~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp:466:55: error: cannot bind non-const lvalue reference of type 'void*&' to an rvalue of type 'void*'
:info:build   466 |                                 char * name = getName(source);
:info:build       |                                                       ^~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp:401:31: note:   initializing argument 1 of 'char* getName(void*&)'
:info:build   401 | char *getName( MIDIObjectRef &object )
:info:build       |                ~~~~~~~~~~~~~~~^~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp:480:55: error: cannot bind non-const lvalue reference of type 'void*&' to an rvalue of type 'void*'
:info:build   480 |                                 char * name = getName(dest);
:info:build       |                                                       ^~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp:401:31: note:   initializing argument 1 of 'char* getName(void*&)'
:info:build   401 | char *getName( MIDIObjectRef &object )
:info:build       |                ~~~~~~~~~~~~~~~^~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp: In static member function 'static void MidiApple::NotifyCallback(const MIDINotification*, void*)':
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp:549:45: warning: format '%d' expects argument of type 'int', but argument 2 has type 'long int' [-Wformat=]
:info:build   549 |         qDebug("MidiApple::NotifyCallback '%d'",message->messageID);
:info:build       |                                            ~^   ~~~~~~~~~~~~~~~~~~
:info:build       |                                             |            |
:info:build       |                                             int          long int
:info:build       |                                            %ld
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp: In member function 'char* MidiApple::getFullName(OpaqueMIDIEndpoint*&)':
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp:615:37: error: cannot bind non-const lvalue reference of type 'void*&' to an rvalue of type 'void*'
:info:build   615 |         char * deviceName = getName(device);
:info:build       |                                     ^~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp:401:31: note:   initializing argument 1 of 'char* getName(void*&)'
:info:build   401 | char *getName( MIDIObjectRef &object )
:info:build       |                ~~~~~~~~~~~~~~~^~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp:616:39: error: cannot bind non-const lvalue reference of type 'void*&' to an rvalue of type 'void*'
:info:build   616 |         char * endPointName = getName(endpoint_ref);
:info:build       |                                       ^~~~~~~~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src/core/midi/MidiApple.cpp:401:31: note:   initializing argument 1 of 'char* getName(void*&)'
:info:build   401 | char *getName( MIDIObjectRef &object )
:info:build       |                ~~~~~~~~~~~~~~~^~~~~~
:info:build make[2]: *** [src/CMakeFiles/lmmsobjs.dir/core/midi/MidiApple.cpp.o] Error 1
:info:build make[2]: *** Waiting for unfinished jobs....
```
Any idea how to fix this?

Given what I have seen in CMakeLists, I assume nobody tested this with GCC on macOS for years, if ever.

@barracuda156 I've simplified your logs by removing the warnings and long path names in an attempt to make them easier to read... 

<details>
<summary>Click to view logs</summary>

```
  [ 56%] Building CXX object  src/CMakeFiles/lmmsobjs.dir/core/midi/MidiPort.cpp.o
  cd /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src && /opt/local/bin/g++-mp-12 -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NO_DEBUG -DQT_XML_LIB -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/  src/lmmsobjs_autogen/include -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/src -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2 -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/include -I/opt/local/include/opus -I/opt/local/include/SDL -I/opt/local/include -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/  src/3rdparty/rpmalloc/rpmalloc/rpmalloc -isystem /opt/local/libexec/qt4/include -isystem /opt/local/libexec/qt4/include/QtGui -isystem /opt/local/libexec/qt4/include/QtXml -isystem /opt/local/libexec/qt4/include/QtCore -fno-exceptions -Wall -Werror=unused-function -Wno-sign-compare -Wno-strict-overflow -Wno-array-bounds -Wno-attributes -pipe -Os -fPIC -DPIC -std=c++0x -DNDEBUG -arch ppc -mmacosx-version-min=10.6   -D'QT_TRANSLATIONS_DIR="/opt/local/libexec/qt4/share/translations"' -D'LIB_DIR="../lib/"' -D'PLUGIN_DIR="../lib/lmms/"' -MD -MT  src/CMakeFiles/lmmsobjs.dir/core/midi/MidiPort.cpp.o -MF CMakeFiles/lmmsobjs.dir/core/midi/MidiPort.cpp.o.d -o CMakeFiles/lmmsobjs.dir/core/midi/MidiPort.cpp.o -c /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_lmms/lmms/work/lmms-1.2.2/  src/core/midi/MidiPort.cpp
  src/core/midi/MidiApple.cpp: In member function 'void MidiApple::openDevices()':
  src/core/midi/MidiApple.cpp:442:45: error: cannot bind non-const lvalue reference of type 'void*&' to an rvalue of type 'void*'
    442 |                 char * deviceName = getName(device);
        |                                             ^~~~~~
  src/core/midi/MidiApple.cpp:401:31: note:   initializing argument 1 of 'char* getName(void*&)'
    401 | char *getName( MIDIObjectRef &object )
        |                ~~~~~~~~~~~~~~~^~~~~~
  src/core/midi/MidiApple.cpp:458:56: error: cannot bind non-const lvalue reference of type 'void*&' to an rvalue of type 'void*'
    458 |                         qDebug("  Entity: %s", getName(entity));
        |                                                        ^~~~~~
  src/core/midi/MidiApple.cpp:401:31: note:   initializing argument 1 of 'char* getName(void*&)'
    401 | char *getName( MIDIObjectRef &object )
        |                ~~~~~~~~~~~~~~~^~~~~~
  src/core/midi/MidiApple.cpp:466:55: error: cannot bind non-const lvalue reference of type 'void*&' to an rvalue of type 'void*'
    466 |                                 char * name = getName(source);
        |                                                       ^~~~~~
  src/core/midi/MidiApple.cpp:401:31: note:   initializing argument 1 of 'char* getName(void*&)'
    401 | char *getName( MIDIObjectRef &object )
        |                ~~~~~~~~~~~~~~~^~~~~~
  src/core/midi/MidiApple.cpp:480:55: error: cannot bind non-const lvalue reference of type 'void*&' to an rvalue of type 'void*'
    480 |                                 char * name = getName(dest);
        |                                                       ^~~~
  src/core/midi/MidiApple.cpp:401:31: note:   initializing argument 1 of 'char* getName(void*&)'
    401 | char *getName( MIDIObjectRef &object )
        |                ~~~~~~~~~~~~~~~^~~~~~
  src/core/midi/MidiApple.cpp: In member function 'char* MidiApple::getFullName(OpaqueMIDIEndpoint*&)':
  src/core/midi/MidiApple.cpp:615:37: error: cannot bind non-const lvalue reference of type 'void*&' to an rvalue of type 'void*'
    615 |         char * deviceName = getName(device);
        |                                     ^~~~~~
  src/core/midi/MidiApple.cpp:401:31: note:   initializing argument 1 of 'char* getName(void*&)'
    401 | char *getName( MIDIObjectRef &object )
        |                ~~~~~~~~~~~~~~~^~~~~~
  src/core/midi/MidiApple.cpp:616:39: error: cannot bind non-const lvalue reference of type 'void*&' to an rvalue of type 'void*'
    616 |         char * endPointName = getName(endpoint_ref);
        |                                       ^~~~~~~~~~~~
  src/core/midi/MidiApple.cpp:401:31: note:   initializing argument 1 of 'char* getName(void*&)'
    401 | char *getName( MIDIObjectRef &object )
        |                ~~~~~~~~~~~~~~~^~~~~~
  make[2]: *** [  src/CMakeFiles/lmmsobjs.dir/core/midi/MidiApple.cpp.o] Error 1
  make[2]: *** Waiting for unfinished jobs....

```

</details>

> @tresf By the way, if you could suggest some reasonable modification to the code to bypass the error, would be very helpful

I'm ironically very poor at C++, just dedicated -- like you -- to get Mac support so I have been heavily involved over the years where I can.

The problem seems to stem from here: 

https://github.com/LMMS/lmms/blob/a9d49d4ff70642c958380a0b2e4fc6180c1a1877/src/core/midi/MidiApple.cpp#L446-L447

From what I can tell, the API wants a [`MIDIObjectRef`](https://developer.apple.com/documentation/coremidi/midiobjectref), but we provide a [`MIDIDeviceRef`](https://developer.apple.com/documentation/coremidi/midideviceref), but I may be wrong.  If it's a conflicting type, you may consider casting it to the correct type.

> Not to be merged, of course, I wanna see if it works locally to begin with.

If the fix works everywhere, we'd keep it.

I'm not sure if you have access to Discord, but posting this on `#dev-support` may be faster way to get it fixed.  Feel free to link to this bug report.  https://lmms.io/chat

--- a/src/core/midi/MidiApple.cpp
+++ b/src/core/midi/MidiApple.cpp
@@ -403,7 +403,7 @@ void MidiApple::midiInClose( MIDIEndpointRef reference )
 
 
 
-char *getName( MIDIObjectRef &object )
+char *getName( const MIDIObjectRef &object )
 {
 	// Returns the name of a given MIDIObjectRef as char *
 	CFStringRef name = nullptr;

